### PR TITLE
Move Polls to CommandsNext

### LINKED ISSUE #14: Polls
- [x] `.abc "question" "option one" "option two" ... "option n"` for creation of multichoice polls
- [x] `.yesno question` for creation of yes no poll
- [ ] Add custom emote for abc type poll

Do we want option to specify emojis in a multichoice poll?

Do we want option to say some form of closing of the polls?

### PR DESCRIPTION
**Pull request by albertmoravec**
Adapt poll command to the CommandsNext interface.

Possible todo:
- [x] sanitize user input

### CODE REVIEW

File: src/HonzaBotner.Discord/Extensions/StringExtension.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+                    DiscordMember member = guild.GetMemberAsync(userId).Result;
+                    return member.RatherNicknameThanUsername();
+                }
+                catch (Exception e)
+                {
+                    // TODO: how to use _logger?
+                    return "";
+                }
+            });
+        }
```
Comment (tenhobi) [inline src/HonzaBotner.Discord/Extensions/StringExtension.cs]: Can we somehow do an async regex replace?
~~How can we use logger in extensions? Should I pass it as an optional parameter?~~

File: src/HonzaBotner.Discord.Services/Commands/PollCommands.cs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+using HonzaBotner.Discord.Extensions;
+using HonzaBotner.Discord.Services.Commands.Polls;
+
+namespace HonzaBotner.Discord.Services.Commands
+{
+    [Group("poll")]
+    [Description("Commands to create polls.")]
+    public class PollCommands : BaseCommandModule
+    {
+        private readonly string _pollErrorMessage =
```
Comment (ostorc) [inline src/HonzaBotner.Discord.Services/Commands/PollCommands.cs]: This should be const
```suggestion
        private const string PollErrorMessage =
```

File: src/HonzaBotner.Discord.Services/Commands/PollCommands.cs [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+            }
+
+            try
+            {
+                await poll.Post(ctx.Client, ctx.Channel);
+                await ctx.Message.DeleteAsync();
+            }
+            catch
+            {
+                await ctx.RespondAsync(_pollErrorMessage);
```
Comment (ostorc) [inline src/HonzaBotner.Discord.Services/Commands/PollCommands.cs]:
```suggestion
                await ctx.RespondAsync(PollErrorMessage);
```

File: src/HonzaBotner.Discord.Services/Commands/PollCommands.cs [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        {
+            try
+            {
+                await new YesNoPoll(ctx.Member.RatherNicknameThanUsername(), ctx.Member.AvatarUrl, question)
+                    .Post(ctx.Client, ctx.Channel);
+                await ctx.Message.DeleteAsync();
+            }
+            catch
+            {
+                await ctx.RespondAsync(_pollErrorMessage);
```
Comment (ostorc) [inline src/HonzaBotner.Discord.Services/Commands/PollCommands.cs]:
```suggestion
                await ctx.RespondAsync(PollErrorMessage);
```

File: src/HonzaBotner.Discord.Services/Commands/PollCommands.cs [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+        {
+            try
+            {
+                await new AbcPoll(ctx.Member.RatherNicknameThanUsername(), ctx.Member.AvatarUrl, question, answers.ToList())
+                    .Post(ctx.Client, ctx.Channel);
+                await ctx.Message.DeleteAsync();
+            }
+            catch
+            {
+                await ctx.RespondAsync(_pollErrorMessage);
```
Comment (ostorc) [inline src/HonzaBotner.Discord.Services/Commands/PollCommands.cs]:
```suggestion
                await ctx.RespondAsync(PollErrorMessage);
```

File: src/HonzaBotner.Discord.Services/Commands/Polls/AbcPoll.cs [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+            // ":regional_indicator_q:",
+            // ":regional_indicator_r:",
+            // ":regional_indicator_s:",
+            // ":regional_indicator_t:",
+            // ":regional_indicator_u:",
+            // ":regional_indicator_v:",
+            // ":regional_indicator_w:",
+            // ":regional_indicator_x:",
+            // ":regional_indicator_y:",
+            // ":regional_indicator_z:"
```
Comment (ostorc) [inline src/HonzaBotner.Discord.Services/Commands/Polls/AbcPoll.cs]:
```suggestion
```
If we dont need them, delete them.

File: src/HonzaBotner.Discord.Services/Commands/Polls/IPoll.cs [hunk 7]
```diff
@@ -0,0 +1,11 @@
+﻿using System.Threading.Tasks;
+using DSharpPlus;
+using DSharpPlus.Entities;
+
+namespace HonzaBotner.Discord.Services.Commands.Polls
+{
+    public interface IPoll
+    {
+        Task Post(DiscordClient client, DiscordChannel channel);
```
Comment (ostorc) [inline src/HonzaBotner.Discord.Services/Commands/Polls/IPoll.cs]: Rename to PostAsync

File: src/HonzaBotner.Discord/Extensions/DiscordMemberExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,12 @@
+﻿using DSharpPlus.Entities;
+
+namespace HonzaBotner.Discord.Extensions
+{
+    public static class DiscordMemberExtensions
+    {
+        public static string RatherNicknameThanUsername(this DiscordMember member)
+        {
+            return member.Nickname ?? member.Username;
```
Comment (ostorc) [inline src/HonzaBotner.Discord/Extensions/DiscordMemberExtensions.cs:9]: Can username be null?

File: src/HonzaBotner.Discord/Extensions/StringExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+using System.Collections.Generic;
+using System.Text.RegularExpressions;
+using DSharpPlus.Entities;
+using Microsoft.Extensions.Logging;
+
+namespace HonzaBotner.Discord.Extensions
+{
+    public static class StringExtension
+    {
+        public static string RemoveDiscordMentions(this string? stringObject, DiscordGuild guild,
```
Comment (ostorc) [inline src/HonzaBotner.Discord/Extensions/StringExtensions.cs]: This whole method is messy, it needs some cleanup

**Review by ostorc — CHANGES_REQUESTED** [review]

File: src/HonzaBotner.Discord/Extensions/DiscordMemberExtensions.cs (hunk 8 above)
Comment (tenhobi) [reply src/HonzaBotner.Discord/Extensions/DiscordMemberExtensions.cs:9]: No.

File: src/HonzaBotner.Discord.Services/Commands/PollCommands.cs [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+
+            try
+            {
+                await poll.PostAsync(ctx.Client, ctx.Channel);
+                await ctx.Message.DeleteAsync();
+            }
+            catch
+            {
+                await ctx.RespondAsync(PollErrorMessage);
+            }
```
Comment (ostorc) [inline src/HonzaBotner.Discord.Services/Commands/PollCommands.cs:54]: Log exception

File: src/HonzaBotner.Discord.Services/Commands/PollCommands.cs [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+            try
+            {
+                await new YesNoPoll(ctx.Member.RatherNicknameThanUsername(), ctx.Member.AvatarUrl, question)
+                    .PostAsync(ctx.Client, ctx.Channel);
+                await ctx.Message.DeleteAsync();
+            }
+            catch
+            {
+                await ctx.RespondAsync(PollErrorMessage);
+            }
```
Comment (ostorc) [inline src/HonzaBotner.Discord.Services/Commands/PollCommands.cs:73]: Log exception

File: src/HonzaBotner.Discord.Services/Commands/PollCommands.cs [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+            try
+            {
+                await new AbcPoll(ctx.Member.RatherNicknameThanUsername(), ctx.Member.AvatarUrl, question, answers.ToList())
+                    .PostAsync(ctx.Client, ctx.Channel);
+                await ctx.Message.DeleteAsync();
+            }
+            catch
+            {
+                await ctx.RespondAsync(PollErrorMessage);
+            }
```
Comment (ostorc) [inline src/HonzaBotner.Discord.Services/Commands/PollCommands.cs:92]: Log exception

**Review by ostorc — CHANGES_REQUESTED** [review]

File: src/HonzaBotner.Discord.Services/Commands/PollCommands.cs (hunk 12 above)
Comment (albertmoravec) [reply src/HonzaBotner.Discord.Services/Commands/PollCommands.cs:92]: Will do once I/someone implements Bot exceptions, so this might wait a bit?

Comment (ostorc) [reply src/HonzaBotner.Discord.Services/Commands/PollCommands.cs:92]: Ok that seems reasonabel

**Review by ostorc — APPROVED** [review]

**Pull request merged**